### PR TITLE
Chat AI assistant

### PR DESCRIPTION
**Pull request by manunam2002**
### Description

This PR introduces an AI assistant that helps the user to write messages in the chat, the user can provide tones for the message, as well as additional info the model should include in the response. The generative model used to achieve this is Google's gemini 1.5 flash.

### Related Issues

Provides additions for #366 

### Changes Made

- Created `ChatAssistantViewModel` to communicate with the generative model
- Created `ChatAssistantDialog`: the ui to prompt the model
- Created tests for these two files

### Demo

https://github.com/user-attachments/assets/9cb03c2a-bd6c-4f1a-9e83-f10612e0303c

### CODE REVIEW

File: app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+          modelName = "gemini-1.5-flash",
+          apiKey = BuildConfig.GOOGLE_AI_API_KEY,
+          generationConfig =
+              generationConfig {
+                temperature = 0.15f
+                topK = 32
+                topP = 1f
+                maxOutputTokens = 4096
+              },
+          safetySettings =
```
Comment (hjamaly) [inline app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt:46]: 👍

File: app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt
Comment (hjamaly) [inline app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt:1]: Excellent Documentation ! Thank you for making the review easier !

File: app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    _senderName.value = ""
+    _receiverName.value = ""
+    _requestContext.value = null
+    _selectedTones.value = emptyList()
+  }
+
+  /**
+   * Build the prompt for the chat assistant
+   *
+   * @param input Additional infos to add to the prompt
```
Comment (hjamaly) [inline app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt:113]: What kind of additional infos ??

File: app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+  /**
+   * Generate a response message from the chat assistant
+   *
+   * @param input Additional infos to add to the prompt
+   * @param onResponse Callback to handle the response
+   */
+  fun generateMessage(input: String, onResponse: (String) -> Unit) {
+    val prompt = buildPrompt(input)
+    Log.d("ChatAssistantViewModel", "Prompt: $prompt")
```
Comment (hjamaly) [inline app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt:151]: Log.d call in generateMessage logs the entire prompt. I'm asking myself if we have to preserve a kind of confidentiality or not ! We can ask the coach about it new meeting !

File: app/src/main/java/com/android/solvit/shared/ui/chat/ChatAssistantDialog.kt
Comment (hjamaly) [inline app/src/main/java/com/android/solvit/shared/ui/chat/ChatAssistantDialog.kt:1]: Good Implementation ! We have to add it on the figma !

File: app/src/main/java/com/android/solvit/shared/ui/chat/ChatScreen.kt [hunk 4]
```diff
@@ -69,13 +72,19 @@ fun ChatScreen(
   val receiverName = getReceiverName(receiver)
   val receiverPicture = getReceiverImageUrl(receiver)
 
+  val user by authViewModel.user.collectAsState()
+  chatAssistantViewModel.setContext(messages, "Hassan", receiverName)
```
Comment (hjamaly) [inline app/src/main/java/com/android/solvit/shared/ui/chat/ChatScreen.kt:76]: Is it normal that the sender name is hard coded ??

File: app/src/test/java/com/android/solvit/shared/model/chat/ChatAssistantViewModelTest.kt
Comment (hjamaly) [inline app/src/test/java/com/android/solvit/shared/model/chat/ChatAssistantViewModelTest.kt:1]: 👍

**Review by hjamaly — APPROVED** [review]
**Feedback** 

- Good Job on implementing this Chat AI Assistant. This introduces an important feature into our App. AI tools are very essential nowadays and it can improve considerably life of user, above all that some user are not necessarily fluent in writing message and that can help them a lot 
- I figured out a little problem. When entering an empty message to generate a response, it gives the previous ai response. Is it desired ? Maybe think to clear the prompt each time a new message is sent
Capture d’écran 2024-11-28 à 22 46 25
Capture d’écran 2024-11-28 à 22 46 30

**Review by aoutz — APPROVED** [review]
This code is a strong and well-structured implementation. The `ChatAssistantViewModel` effectively manages state and handles AI-driven logic. The prompt-building logic is well-designed, dynamically incorporating the conversation context, tones, and user inputs to ensure relevant AI responses. The composables, such as `ToneItem` and `SentMessage`, are thoughtfully crafted, enhancing the user experience with intuitive and visually appealing elements. 

Overall, this is an excellent implementation Manu, and while minor enhancements like better error handling or loading indicators could refine the user experience, the code already stands out as a well-executed and forward-thinking solution.  Chapeau 💪!

File: app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt (hunk 2 above)
Comment (manunam2002) [reply app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt:113]: this is for whatever additional info we may want to add into the response

File: app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt (hunk 3 above)
Comment (manunam2002) [reply app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt:151]: I can remove it!

File: app/src/main/java/com/android/solvit/shared/ui/chat/ChatScreen.kt (hunk 4 above)
Comment (manunam2002) [reply app/src/main/java/com/android/solvit/shared/ui/chat/ChatScreen.kt:76]: It is not, but it is hardcoded also in the text messages, so I could not make better!

File: app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    return prompt
+  }
+
+  /**
+   * Generate a response message from the chat assistant
+   *
+   * @param input Additional infos to add to the prompt
+   * @param onResponse Callback to handle the response
+   */
+  fun generateMessage(input: String, onResponse: (String) -> Unit) {
```
[the remark this answers is not in the document]
Comment (manunam2002) [reply app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt:149]: thanks a lot, great suggestion!

File: app/src/main/java/com/android/solvit/shared/ui/chat/ChatScreen.kt (hunk 4 above)
Comment (hjamaly) [reply app/src/main/java/com/android/solvit/shared/ui/chat/ChatScreen.kt:76]: Yes Indeed! I'll make a PR to correct it !

File: app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    val prompt = buildPrompt(input)
+    Log.d("ChatAssistantViewModel", "Prompt: $prompt")
+    viewModelScope.launch {
+      val response = model.generateContent(prompt)
```
[the remark this answers is not in the document]
Comment (manunam2002) [reply app/src/main/java/com/android/solvit/shared/model/chat/ChatAssistantViewModel.kt]: done, thanks for the suggestion!

**Pull request merged**